File: src/symbolic_math/physics.py
"""General relativity and field theory computations.

Computes Christoffel symbols, Riemann tensor, Ricci tensor/scalar,
and Einstein tensor from a given metric.
"""

from __future__ import annotations

from typing import Optional

import sympy
from sympy import Matrix, Symbol, simplify, zeros, latex, Rational


def _parse_metric(metric: list[list[str]], coords: list[str]) -> tuple[Matrix, list[Symbol]]:
    """Parse metric matrix and coordinate list into SymPy objects."""
    coord_syms = [Symbol(c) for c in coords]
    n = len(coords)
    g = zeros(n)
    for i in range(n):
        for j in range(n):
            g[i, j] = sympy.sympify(metric[i][j])
    return g, coord_syms


def _metric_inverse(g: Matrix) -> Matrix:
    """Compute the inverse metric tensor."""
    return g.inv()


def christoffel(metric: list[list[str]], coords: list[str]) -> dict:
    """Compute Christoffel symbols of the second kind.

    Returns Gamma^sigma_{mu nu} as a nested dict of non-zero components.
    """
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)
    n = len(coords)

    gamma = {}
    components = {}

    for sigma in range(n):
        for mu in range(n):
            for nu in range(mu, n):  # symmetric in lower indices
                val = Rational(0)
                for rho in range(n):
                    val += Rational(1, 2) * g_inv[sigma, rho] * (
                        sympy.diff(g[nu, rho], coord_syms[mu])
                        + sympy.diff(g[mu, rho], coord_syms[nu])
                        - sympy.diff(g[mu, nu], coord_syms[rho])
                    )
                val = simplify(val)
                if val != 0:
                    key = f"Gamma^{coords[sigma]}_{{{coords[mu]}{coords[nu]}}}"
                    components[key] = {
                        "value": str(val),
                        "latex": latex(val),
                    }

    return {
        "result": components if components else "All Christoffel symbols are zero",
        "latex": _christoffel_latex(components) if components else "\\Gamma^{\\sigma}_{\\mu\\nu} = 0",
        "type": "christoffel_symbols",
    }


def _christoffel_latex(components: dict) -> str:
    """Format Christoffel symbols as LaTeX."""
    lines = []
    for key, val in components.items():
        lines.append(f"{key} = {val['latex']}")
    return " \\\\\n".join(lines)


def riemann_tensor(metric: list[list[str]], coords: list[str]) -> dict:
    """Compute the Riemann curvature tensor R^rho_{sigma mu nu}.

    Returns non-zero components.
    """
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)
    n = len(coords)

    # First compute Christoffel symbols as a 3D array
    gamma = _christoffel_array(g, g_inv, coord_syms)

    components = {}
    for rho in range(n):
        for sigma in range(n):
            for mu in range(n):
                for nu in range(mu + 1, n):  # antisymmetric in mu,nu
                    val = (
                        sympy.diff(gamma[rho][nu][sigma], coord_syms[mu])
                        - sympy.diff(gamma[rho][mu][sigma], coord_syms[nu])
                    )
                    for lam in range(n):
                        val += (
                            gamma[rho][mu][lam] * gamma[lam][nu][sigma]
                            - gamma[rho][nu][lam] * gamma[lam][mu][sigma]
                        )
                    val = simplify(val)
                    if val != 0:
                        key = f"R^{coords[rho]}_{{{coords[sigma]}{coords[mu]}{coords[nu]}}}"
                        components[key] = {
                            "value": str(val),
                            "latex": latex(val),
                        }

    return {
        "result": components if components else "Riemann tensor is identically zero (flat space)",
        "latex": _format_tensor_latex(components, "R") if components else "R^{\\rho}_{\\sigma\\mu\\nu} = 0",
        "type": "riemann_tensor",
    }


def ricci_tensor(metric: list[list[str]], coords: list[str]) -> dict:
    """Compute the Ricci tensor R_{mu nu} by contracting the Riemann tensor."""
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)
    n = len(coords)

    gamma = _christoffel_array(g, g_inv, coord_syms)
    riemann = _riemann_array(gamma, coord_syms, n)

    components = {}
    ricci_matrix = zeros(n)
    for mu in range(n):
        for nu in range(mu, n):  # symmetric
            val = Rational(0)
            for rho in range(n):
                val += riemann[rho][mu][rho][nu]
            val = simplify(val)
            ricci_matrix[mu, nu] = val
            ricci_matrix[nu, mu] = val
            if val != 0:
                key = f"R_{{{coords[mu]}{coords[nu]}}}"
                components[key] = {
                    "value": str(val),
                    "latex": latex(val),
                }

    return {
        "result": components if components else "Ricci tensor is identically zero",
        "latex": _format_tensor_latex(components, "R") if components else "R_{\\mu\\nu} = 0",
        "type": "ricci_tensor",
        "matrix": [[str(ricci_matrix[i, j]) for j in range(n)] for i in range(n)],
    }


def ricci_scalar(metric: list[list[str]], coords: list[str]) -> dict:
    """Compute the Ricci scalar R = g^{mu nu} R_{mu nu}."""
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)
    n = len(coords)

    gamma = _christoffel_array(g, g_inv, coord_syms)
    riemann = _riemann_array(gamma, coord_syms, n)

    # Ricci tensor
    ricci = zeros(n)
    for mu in range(n):
        for nu in range(n):
            val = Rational(0)
            for rho in range(n):
                val += riemann[rho][mu][rho][nu]
            ricci[mu, nu] = val

    # Contract with inverse metric
    scalar = Rational(0)
    for mu in range(n):
        for nu in range(n):
            scalar += g_inv[mu, nu] * ricci[mu, nu]
    scalar = simplify(scalar)

    return {
        "result": str(scalar),
        "latex": latex(scalar),
        "type": "scalar",
    }


def einstein_tensor(metric: list[list[str]], coords: list[str]) -> dict:
    """Compute the Einstein tensor G_{mu nu} = R_{mu nu} - 1/2 R g_{mu nu}."""
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)
    n = len(coords)

    gamma = _christoffel_array(g, g_inv, coord_syms)
    riemann = _riemann_array(gamma, coord_syms, n)

    # Ricci tensor
    ricci = zeros(n)
    for mu in range(n):
        for nu in range(n):
            val = Rational(0)
            for rho in range(n):
                val += riemann[rho][mu][rho][nu]
            ricci[mu, nu] = val

    # Ricci scalar
    scalar = Rational(0)
    for mu in range(n):
        for nu in range(n):
            scalar += g_inv[mu, nu] * ricci[mu, nu]
    scalar = simplify(scalar)

    # Einstein tensor
    components = {}
    einstein_matrix = zeros(n)
    for mu in range(n):
        for nu in range(mu, n):
            val = simplify(ricci[mu, nu] - Rational(1, 2) * scalar * g[mu, nu])
            einstein_matrix[mu, nu] = val
            einstein_matrix[nu, mu] = val
            if val != 0:
                key = f"G_{{{coords[mu]}{coords[nu]}}}"
                components[key] = {
                    "value": str(val),
                    "latex": latex(val),
                }

    return {
        "result": components if components else "Einstein tensor is identically zero (vacuum solution)",
        "latex": _format_tensor_latex(components, "G") if components else "G_{\\mu\\nu} = 0",
        "type": "einstein_tensor",
        "matrix": [[str(einstein_matrix[i, j]) for j in range(n)] for i in range(n)],
    }


def covariant_derivative(expr_str: str, metric: list[list[str]],
                         coords: list[str], direction: str) -> dict:
    """Compute the covariant derivative of a scalar expression."""
    g, coord_syms = _parse_metric(metric, coords)
    g_inv = _metric_inverse(g)

    expr = sympy.sympify(expr_str)
    dir_idx = coords.index(direction)

    # For a scalar, covariant derivative = partial derivative
    result = simplify(sympy.diff(expr, coord_syms[dir_idx]))

    return {
        "result": str(result),
        "latex": latex(result),
        "type": "expression",
    }


def evaluate_component(expr_str: str, substitutions: dict) -> dict:
    """Evaluate a tensor component expression numerically."""
    expr = sympy.sympify(expr_str)
    subs = {Symbol(k): sympy.sympify(v) for k, v in substitutions.items()}
    result = simplify(expr.subs(subs))
    return {
        "result": str(result),
        "latex": latex(result),
        "type": "expression",
    }


# --- Internal helpers ---

def _christoffel_array(g: Matrix, g_inv: Matrix, coords: list[Symbol]) -> list:
    """Compute Christoffel symbols as a 3D nested list."""
    n = len(coords)
    gamma = [[[Rational(0) for _ in range(n)] for _ in range(n)] for _ in range(n)]
    for sigma in range(n):
        for mu in range(n):
            for nu in range(n):
                val = Rational(0)
                for rho in range(n):
                    val += Rational(1, 2) * g_inv[sigma, rho] * (
                        sympy.diff(g[nu, rho], coords[mu])
                        + sympy.diff(g[mu, rho], coords[nu])
                        - sympy.diff(g[mu, nu], coords[rho])
                    )
                gamma[sigma][mu][nu] = simplify(val)
    return gamma


def _riemann_array(gamma: list, coords: list[Symbol], n: int) -> list:
    """Compute Riemann tensor as a 4D nested list."""
    riemann = [[[[Rational(0) for _ in range(n)] for _ in range(n)]
                for _ in range(n)] for _ in range(n)]
    for rho in range(n):
        for sigma in range(n):
            for mu in range(n):
                for nu in range(n):
                    val = (
                        sympy.diff(gamma[rho][nu][sigma], coords[mu])
                        - sympy.diff(gamma[rho][mu][sigma], coords[nu])
                    )
                    for lam in range(n):
                        val += (
                            gamma[rho][mu][lam] * gamma[lam][nu][sigma]
                            - gamma[rho][nu][lam] * gamma[lam][mu][sigma]
                        )
                    riemann[rho][sigma][mu][nu] = simplify(val)
    return riemann


def _format_tensor_latex(components: dict, symbol: str) -> str:
    """Format tensor components as LaTeX."""
    lines = []
    for key, val in components.items():
        lines.append(f"{key} = {val['latex']}")
    return " \\\\\n".join(lines)

File: cpp/src/physics/physics.cpp
#include "physics/physics.h"
#include "physics/metric.h"
#include "engine/parser.h"
#include "engine/output.h"

#include <symengine/expression.h>
#include <symengine/add.h>
#include <symengine/mul.h>
#include <symengine/rational.h>
#include <symengine/constants.h>
#include <symengine/subs.h>
#include <symengine/printers.h>

#include <sstream>

using namespace SymEngine;

namespace physics {

// --- Internal helpers ---

using Tensor3 = std::vector<std::vector<std::vector<RCP<const Basic>>>>;
using Tensor4 = std::vector<std::vector<std::vector<std::vector<RCP<const Basic>>>>>;

static Tensor3 compute_christoffel_array(const MetricData& m) {
    int n = m.n;
    Tensor3 gamma(n, std::vector<std::vector<RCP<const Basic>>>(
        n, std::vector<RCP<const Basic>>(n, zero)));

    for (int sigma = 0; sigma < n; ++sigma) {
        for (int mu = 0; mu < n; ++mu) {
            for (int nu = 0; nu < n; ++nu) {
                RCP<const Basic> val = zero;
                for (int rho = 0; rho < n; ++rho) {
                    auto g_inv_sr = m.g_inv.get(sigma, rho);
                    auto dg_nu_rho = m.g.get(nu, rho)->diff(m.coords[mu]);
                    auto dg_mu_rho = m.g.get(mu, rho)->diff(m.coords[nu]);
                    auto dg_mu_nu  = m.g.get(mu, nu)->diff(m.coords[rho]);

                    val = add(val, mul(
                        div(one, integer(2)),
                        mul(g_inv_sr, add(dg_nu_rho, sub(dg_mu_rho, dg_mu_nu)))
                    ));
                }
                gamma[sigma][mu][nu] = expand(val);
            }
        }
    }
    return gamma;
}

static Tensor4 compute_riemann_array(const Tensor3& gamma,
                                      const std::vector<RCP<const Symbol>>& coords,
                                      int n) {
    Tensor4 riemann(n, Tensor3(n, std::vector<std::vector<RCP<const Basic>>>(
        n, std::vector<RCP<const Basic>>(n, zero))));

    for (int rho = 0; rho < n; ++rho) {
        for (int sigma = 0; sigma < n; ++sigma) {
            for (int mu = 0; mu < n; ++mu) {
                for (int nu = 0; nu < n; ++nu) {
                    // R^rho_{sigma mu nu} = d_mu Gamma^rho_{nu sigma}
                    //                     - d_nu Gamma^rho_{mu sigma}
                    //                     + Gamma^rho_{mu lam} Gamma^lam_{nu sigma}
                    //                     - Gamma^rho_{nu lam} Gamma^lam_{mu sigma}
                    auto val = sub(
                        gamma[rho][nu][sigma]->diff(coords[mu]),
                        gamma[rho][mu][sigma]->diff(coords[nu])
                    );
                    for (int lam = 0; lam < n; ++lam) {
                        val = add(val, sub(
                            mul(gamma[rho][mu][lam], gamma[lam][nu][sigma]),
                            mul(gamma[rho][nu][lam], gamma[lam][mu][sigma])
                        ));
                    }
                    riemann[rho][sigma][mu][nu] = expand(val);
                }
            }
        }
    }
    return riemann;
}

static std::string format_tensor_latex(const json& components) {
    std::string result;
    bool first = true;
    for (auto& [key, val] : components.items()) {
        if (!first) result += " \\\\\n";
        result += key + " = " + val["latex"].get<std::string>();
        first = false;
    }
    return result;
}

// --- Public API ---

json christoffel(const std::vector<std::vector<std::string>>& metric,
                 const std::vector<std::string>& coords) {
    auto m = parse_metric(metric, coords);
    auto gamma = compute_christoffel_array(m);

    json components = json::object();
    for (int sigma = 0; sigma < m.n; ++sigma) {
        for (int mu = 0; mu < m.n; ++mu) {
            for (int nu = mu; nu < m.n; ++nu) {  // symmetric in lower indices
                auto val = gamma[sigma][mu][nu];
                if (!eq(*val, *zero)) {
                    std::string key = "Gamma^" + coords[sigma] + "_{" +
                                      coords[mu] + coords[nu] + "}";
                    components[key] = {
                        {"value", val->__str__()},
                        {"latex", latex(*val)},
                    };
                }
            }
        }
    }

    if (components.empty()) {
        return {
            {"result", "All Christoffel symbols are zero"},
            {"latex", "\\Gamma^{\\sigma}_{\\mu\\nu} = 0"},
            {"type", "christoffel_symbols"},
        };
    }

    return {
        {"result", components},
        {"latex", format_tensor_latex(components)},
        {"type", "christoffel_symbols"},
    };
}

json riemann_tensor(const std::vector<std::vector<std::string>>& metric,
                    const std::vector<std::string>& coords) {
    auto m = parse_metric(metric, coords);
    auto gamma = compute_christoffel_array(m);
    auto riemann = compute_riemann_array(gamma, m.coords, m.n);

    json components = json::object();
    for (int rho = 0; rho < m.n; ++rho) {
        for (int sigma = 0; sigma < m.n; ++sigma) {
            for (int mu = 0; mu < m.n; ++mu) {
                for (int nu = mu + 1; nu < m.n; ++nu) {  // antisymmetric
                    auto val = riemann[rho][sigma][mu][nu];
                    if (!eq(*val, *zero)) {
                        std::string key = "R^" + coords[rho] + "_{" +
                                          coords[sigma] + coords[mu] + coords[nu] + "}";
                        components[key] = {
                            {"value", val->__str__()},
                            {"latex", latex(*val)},
                        };
                    }
                }
            }
        }
    }

    if (components.empty()) {
        return {
            {"result", "Riemann tensor is identically zero (flat space)"},
            {"latex", "R^{\\rho}_{\\sigma\\mu\\nu} = 0"},
            {"type", "riemann_tensor"},
        };
    }

    return {
        {"result", components},
        {"latex", format_tensor_latex(components)},
        {"type", "riemann_tensor"},
    };
}

json ricci_tensor(const std::vector<std::vector<std::string>>& metric,
                  const std::vector<std::string>& coords) {
    auto m = parse_metric(metric, coords);
    auto gamma = compute_christoffel_array(m);
    auto riemann = compute_riemann_array(gamma, m.coords, m.n);

    json components = json::object();
    // Ricci matrix for output
    std::vector<std::vector<std::string>> ricci_matrix(m.n,
        std::vector<std::string>(m.n, "0"));

    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = mu; nu < m.n; ++nu) {  // symmetric
            RCP<const Basic> val = zero;
            for (int rho = 0; rho < m.n; ++rho) {
                val = add(val, riemann[rho][mu][rho][nu]);
            }
            val = expand(val);
            ricci_matrix[mu][nu] = val->__str__();
            ricci_matrix[nu][mu] = val->__str__();
            if (!eq(*val, *zero)) {
                std::string key = "R_{" + coords[mu] + coords[nu] + "}";
                components[key] = {
                    {"value", val->__str__()},
                    {"latex", latex(*val)},
                };
            }
        }
    }

    json result;
    if (components.empty()) {
        result = {
            {"result", "Ricci tensor is identically zero"},
            {"latex", "R_{\\mu\\nu} = 0"},
            {"type", "ricci_tensor"},
        };
    } else {
        result = {
            {"result", components},
            {"latex", format_tensor_latex(components)},
            {"type", "ricci_tensor"},
        };
    }
    result["matrix"] = ricci_matrix;
    return result;
}

json ricci_scalar(const std::vector<std::vector<std::string>>& metric,
                  const std::vector<std::string>& coords) {
    auto m = parse_metric(metric, coords);
    auto gamma = compute_christoffel_array(m);
    auto riemann = compute_riemann_array(gamma, m.coords, m.n);

    // Ricci tensor R_{mu nu}
    DenseMatrix ricci(m.n, m.n);
    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = 0; nu < m.n; ++nu) {
            RCP<const Basic> val = zero;
            for (int rho = 0; rho < m.n; ++rho) {
                val = add(val, riemann[rho][mu][rho][nu]);
            }
            ricci.set(mu, nu, val);
        }
    }

    // Contract: R = g^{mu nu} R_{mu nu}
    RCP<const Basic> scalar = zero;
    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = 0; nu < m.n; ++nu) {
            scalar = add(scalar, mul(m.g_inv.get(mu, nu), ricci.get(mu, nu)));
        }
    }
    scalar = expand(scalar);

    return {
        {"result", scalar->__str__()},
        {"latex", latex(*scalar)},
        {"type", "scalar"},
    };
}

json einstein_tensor(const std::vector<std::vector<std::string>>& metric,
                     const std::vector<std::string>& coords) {
    auto m = parse_metric(metric, coords);
    auto gamma = compute_christoffel_array(m);
    auto riemann = compute_riemann_array(gamma, m.coords, m.n);

    // Ricci tensor
    DenseMatrix ricci(m.n, m.n);
    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = 0; nu < m.n; ++nu) {
            RCP<const Basic> val = zero;
            for (int rho = 0; rho < m.n; ++rho) {
                val = add(val, riemann[rho][mu][rho][nu]);
            }
            ricci.set(mu, nu, val);
        }
    }

    // Ricci scalar
    RCP<const Basic> scalar = zero;
    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = 0; nu < m.n; ++nu) {
            scalar = add(scalar, mul(m.g_inv.get(mu, nu), ricci.get(mu, nu)));
        }
    }
    scalar = expand(scalar);

    // G_{mu nu} = R_{mu nu} - 1/2 R g_{mu nu}
    json components = json::object();
    std::vector<std::vector<std::string>> einstein_matrix(m.n,
        std::vector<std::string>(m.n, "0"));

    for (int mu = 0; mu < m.n; ++mu) {
        for (int nu = mu; nu < m.n; ++nu) {
            auto val = expand(sub(ricci.get(mu, nu),
                                   mul(div(one, integer(2)),
                                       mul(scalar, m.g.get(mu, nu)))));
            einstein_matrix[mu][nu] = val->__str__();
            einstein_matrix[nu][mu] = val->__str__();
            if (!eq(*val, *zero)) {
                std::string key = "G_{" + coords[mu] + coords[nu] + "}";
                components[key] = {
                    {"value", val->__str__()},
                    {"latex", latex(*val)},
                };
            }
        }
    }

    json result;
    if (components.empty()) {
        result = {
            {"result", "Einstein tensor is identically zero (vacuum solution)"},
            {"latex", "G_{\\mu\\nu} = 0"},
            {"type", "einstein_tensor"},
        };
    } else {
        result = {
            {"result", components},
            {"latex", format_tensor_latex(components)},
            {"type", "einstein_tensor"},
        };
    }
    result["matrix"] = einstein_matrix;
    return result;
}

json covariant_derivative(const std::string& expr_str,
                          const std::vector<std::vector<std::string>>& metric,
                          const std::vector<std::string>& coords,
                          const std::string& direction) {
    auto m = parse_metric(metric, coords);
    auto expr = engine::parse_expr(expr_str);

    // Find direction index
    int dir_idx = -1;
    for (int i = 0; i < m.n; ++i) {
        if (coords[i] == direction) {
            dir_idx = i;
            break;
        }
    }
    if (dir_idx < 0) {
        throw std::runtime_error("Direction '" + direction + "' not in coordinates");
    }

    // For a scalar, covariant derivative = partial derivative
    auto result = expr.get_basic()->diff(m.coords[dir_idx]);
    result = expand(result);

    return engine::to_output(Expression(result));
}

json evaluate_component(const std::string& expr_str,
                        const std::map<std::string, std::string>& substitutions) {
    auto expr = engine::parse_expr(expr_str);
    auto basic = expr.get_basic();

    map_basic_basic sub_map;
    for (const auto& [k, v] : substitutions) {
        sub_map[symbol(k)] = engine::parse_expr(v).get_basic();
    }

    auto result = basic->subs(sub_map);
    result = expand(result);
    return engine::to_output(Expression(result));
}

} // namespace physics
